include/deal.II/fe/fe_coupling_values.h

```diff
@@ -0,0 +1,653 @@
+        return view.value(inner_shape_function, inner_q_point);
+      }
+  }
+
+
+
+  template <typename ViewType>
+  typename RenumberedView<ViewType>::gradient_type
+  RenumberedView<ViewType>::gradient(const unsigned int shape_function,
+                                     const unsigned int q_point) const
+  {
+    AssertIndexRange(shape_function, data.n_dofs);
+    AssertIndexRange(q_point, data.n_quadrature_points);
+
+    const auto inner_shape_function = data.dof_renumbering.empty() ?
+                                        shape_function :
+                                        data.dof_renumbering[shape_function];
+    const auto inner_q_point        = data.quadrature_renumbering.empty() ?
+                                        q_point :
+                                        data.quadrature_renumbering[q_point];
+    if (inner_shape_function == numbers::invalid_unsigned_int)
+      return gradient_type();
+    else
+      return view.gradient(inner_shape_function, inner_q_point);
+  }
+
+
+
+  template <typename ViewType>
+  template <typename ValueType>
+  std::vector<ValueType> &
+  RenumberedView<ViewType>::outer_to_inner_values(
+    std::vector<ValueType> &outer_values) const
+  {
+    AssertDimension(outer_values.size(), data.n_quadrature_points);
+    if (data.quadrature_renumbering.empty())
+      {
+        return outer_values;
+      }
+    else
+      {
+        const auto name = data.get_unique_container_name(
+          "RenumberedView::outer_to_inner_values",
+          data.n_inner_quadrature_points,
+          outer_values[0]);
+        auto &inner_values =
+          data.data_storage.get()
+            .template get_or_add_object_with_name<std::vector<ValueType>>(
+              name, data.n_inner_quadrature_points);
+        return inner_values;
+      }
+  }
+
+
+
+  template <typename ViewType>
+  template <typename VectorType>
+  const VectorType &
+  RenumberedView<ViewType>::outer_to_inner_dofs(
+    const VectorType &outer_dofs) const
+  {
+    AssertDimension(outer_dofs.size(), data.n_dofs);
+    if (data.dof_renumbering.empty())
+      {
+        return outer_dofs;
+      }
+    else
+      {
+        const auto name =
+          data.get_unique_container_name("RenumberedView::outer_to_inner_dofs",
+                                         data.n_inner_dofs,
+                                         outer_dofs[0]);
+        auto &inner_dofs = data.data_storage.get()
+                             .template get_or_add_object_with_name<VectorType>(
+                               name, data.n_inner_dofs);
+        for (unsigned int i = 0; i < data.n_dofs; ++i)
+          {
+            const auto inner_i = data.dof_renumbering[i];
+            if (inner_i != numbers::invalid_unsigned_int)
+              {
+                AssertIndexRange(inner_i, data.n_inner_dofs);
+                inner_dofs[inner_i] = outer_dofs[i];
+              }
+          }
```

Comment:
Here, if you get to this function a second time, don't you want to skip the initialization of the `inner_dofs` array?

Reply:
No. Calling this function is done to translate outer degrees of freedom into inner degrees of freedom before calling `get_function_values_from_dof_values(..., inner_dofs)`, so it needs the renumbered dofs passed as arguments.

Reply:
Oh, I had missed that you're already doing the copying from `outer_dofs` here. These will be different every time, so the code block doesn't just initialize the object, but actually do the work on it.